File: geocoding/src/main/java/se/arbitur/geocoding/Result.java
package se.arbitur.geocoding;


import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

import se.arbitur.geocoding.Models.Coordinate;



public final class Result {
	@SerializedName("address_components") AddressComponent[] addressComponents;
	@SerializedName("formatted_address") String formattedAddress;
	@SerializedName("geometry") Geometry geometry;
	@SerializedName("types") String[] addressTypes;
	@SerializedName("place_id") String placeId;


	public Geometry getGeometry() {
		return geometry;
	}

	public AddressComponent[] getAddressComponents() {
		return addressComponents;
	}

	public String[] getAddressTypes() {
		return addressTypes;
	}

	public String getFormattedAddress() {
		return formattedAddress;
	}

	public String getPlaceId() {
		return placeId;
	}

	public AddressComponent[] getAddressComponents(String addressType) {
		List<AddressComponent> filter = new ArrayList<>();

		for (AddressComponent component : getAddressComponents())
			for (int i = 0; i < component.getAddressTypes().length; i++) {
				String type = component.getAddressTypes()[i];
				if (type.equals(addressType)) {
					filter.add(component);
					break;
				}
			}

		return filter.toArray(new AddressComponent[filter.size()]);
	}





	public static final class AddressComponent {
		@SerializedName("long_name") String longName;
		@SerializedName("short_name") String shortName;
		@SerializedName("types") String[] addressTypes;


		public String[] getAddressTypes() {
			return addressTypes;
		}

		public String getLongName() {
			return longName;
		}

		public String getShortName() {
			return shortName;
		}
	}





	public static final class Geometry {
		@SerializedName("location") Coordinate location;
		@SerializedName("location_type") String locationType;
		@SerializedName("viewport") Viewport viewport;


		public Coordinate getLocation() {
			return location;
		}

		public String getLocationType() {
			return locationType;
		}

		public Viewport getViewport() {
			return viewport;
		}




		public static final class Viewport {
			@SerializedName("southwest") Coordinate southWest;
			@SerializedName("northeast") Coordinate northEast;


			public Coordinate getNorthEast() {
				return northEast;
			}

			public Coordinate getSouthWest() {
				return southWest;
			}
		}
	}
}

File: geocoding/src/main/java/se/arbitur/geocoding/Result.kt
package se.arbitur.geocoding


import org.json.JSONObject
import se.arbitur.geocoding.models.Coordinate



class Result(json: JSONObject) {

	val addressComponents: Array<AddressComponent>
	val formattedAddress: String = json.getString("formatted_address")
	val geometry: Geometry = Geometry(json.getJSONObject("geometry"))
	val addressTypes: Array<String>
	val placeId: String = json.getString("place_id")



	init {
		val addressComponents = json.getJSONArray("address_components")
		this.addressComponents = (0 until addressComponents.length()).map { AddressComponent(addressComponents.getJSONObject(it)) }.toTypedArray()

		val addressTypes = json.getJSONArray("types")
		this.addressTypes = (0 until addressTypes.length()).map { addressTypes.getString(it) }.toTypedArray()
	}



	fun filterAddressComponents(type: String): Array<AddressComponent> = addressComponents.filter {
			it.addressTypes.contains(type)
		}.toTypedArray()



	class AddressComponent(json: JSONObject) {
		val longName: String = json.getString("long_name")
		val shortName: String = json.getString("short_name")
		val addressTypes: Array<String>

		init {
			val addressTypes = json.getJSONArray("types")
			this.addressTypes = (0 until addressTypes.length()).map { addressTypes.getString(it) }.toTypedArray()
		}
	}



	class Geometry(json: JSONObject) {
		val location: Coordinate
		val locationType: String = json.getString("location_type")
		val viewport: Viewport = Viewport(json.getJSONObject("viewport"))

		init {
			val jsonLocation = json.getJSONObject("location")
			location = Coordinate(jsonLocation.getDouble("lat"), jsonLocation.getDouble("lng"))
		}

		class Viewport(json: JSONObject) {
			val southWest: Coordinate
			val northEast: Coordinate

			init {
				val jsonSW = json.getJSONObject("southwest")
				southWest = Coordinate(jsonSW.getDouble("lat"), jsonSW.getDouble("lng"))

				val jsonNE = json.getJSONObject("northeast")
				northEast = Coordinate(jsonNE.getDouble("lat"), jsonNE.getDouble("lng"))
			}
		}
	}
}
